python/cuml/test/test_kmeans.py

```diff
@@ -109,6 +109,45 @@ def test_kmeans_sequential_plus_plus_init(nrows, ncols, nclusters,
     assert abs(cu_score - sk_score) <= cluster_std * 1.5
 
 
+@pytest.mark.parametrize('nrows', [1000, 10000])
+@pytest.mark.parametrize('ncols', [10, 50])
+@pytest.mark.parametrize('nclusters', [5, 10])
+@pytest.mark.parametrize('max_weight', [10, 20])
+@pytest.mark.parametrize('random_state', [i for i in range(5)])
+def test_weighted_kmeans(nrows, ncols, nclusters,
+                         max_weight, random_state):
+
+    # Using fairly high variance between points in clusters
+    cluster_std = 1.0
+    np.random.seed(random_state)
+
+    # set weight per sample to be from 1 to max_weight
+    wt = np.random.randint(1, high=max_weight, size=nrows)
+
+    X, y = make_blobs(nrows,
+                      ncols,
+                      nclusters,
+                      cluster_std=cluster_std,
+                      shuffle=False,
+                      random_state=0)
+
+    cuml_kmeans = cuml.KMeans(init="k-means++",
+                              n_clusters=nclusters,
+                              n_init=10,
+                              random_state=random_state,
+                              output_type='numpy')
+
+    cuml_kmeans.fit(X, sample_weight=wt)
+    cu_score = cuml_kmeans.score(X)
+
+    sk_kmeans = cluster.KMeans(random_state=random_state,
+                               n_clusters=nclusters)
+    sk_kmeans.fit(X.copy_to_host(), sample_weight=wt)
+    sk_score = sk_kmeans.score(X.copy_to_host())
+
+    assert abs(cu_score - sk_score) <= cluster_std * 1.5
```

Comment:
I know we have a ton of tests like this in the codebase already, but think we've seen that they don't necessarily catch all the corner cases that well, especially when we have to apply these threshold gaps (like 1.5x). It's worth thinking a bit more about some tests (could be on the c++ level instead) to very directly test that weights are working as expected.

E.g. sklearn has a test that repeating one point N times or giving that same point with weight=N should yield identical results. Could be a nice addition.

Reply:
I agree with you @JohnZed. I implemented weighted kmeans unit tests based on how other python tests were implemented and didn't think about focused tests. 
Maybe we can add more focused tests as a part of follow-on PR with @cjnolet's help as he had written most of the python unit tests in kmeans. I had discussed this in the past about refactoring kmeans unit test to make it more concrete because using `assert abs(cu_score - sk_score) <= cluster_std * 1.5` or other similar comparisons against sklearn's score is not a good way to validate the code.

Reply:
> as he had written most of the python unit tests in kmeans.

@akkamesh @JohnZed this statement isn't quite true, in fact, most of the existing Python tests in k-means predate my time on cuML.

I wrote the specific test that this has been copied from, though the original was built for a slightly different purpose. Specifically, this tests the random_state of kmeans with the traditional k-means++ option using `n_init=10` because I found that when we initialize in a similar way to sklearn, we end up with very close WCSSEs. The 1.5 was just meant as a margin of error, since they are not going to be exact, but my original test is executed over multiple random states to demonstrate that the settings are consistent within that tolerance. I used a scaling factor of the cluster std deviation rather than an exact tolerance because I was finding that the differences are proportional to the size of the variance in the input clusters.

@akkamesh, I thinyour test might want to support a different evaluation criteria to verify the weighting is working as expected. For example, creating a skewed distribution of weights and verifying the resulting cluster centers followed that distribution, vs the points having equal weight.

Reply:
Thanks for the clarification @cjnolet. With so many contributions from your end on the python side of k-means, I had assumed that the original unit tests were contributed by you too. I apologize for that.

> I thinyour test might want to support a different evaluation criteria to verify the weighting is working as expected. I think comparing against sklearn could help, but there are other parametrizations that would be helpful as well

Agreed. Need some thinking here. But reproducing the weighted tests from sklearn might be a good place to start.